Integrate with uae-dap 1.x with new patched FS-UAE 4

grahambates opened the pull request:
UAE-DAP has been updated to use a [new remote debugger implementation for FS-UAE](https://github.com/grahambates/fs-uae/tree/barto_prb), which ports the [WinUAE implementation](https://github.com/prb28/WinUAE) to the current dev (4.x) build. This means no longer having to maintain two separate GDB implementations and we're no longer stuck on an old 3.x branch. This has also made a big difference to the complexity and maintainability of the adapter implementation.

This does mean we're on a pre-release version of FS-UAE, but it's totally usable and doesn't seem to be missing any features that are important to our use case. [More details here](http://eab.abime.net/showthread.php?t=106562).

I've done some testing across Mac/Win/Linux and so far it's looking good. One clear benefit is improved startup times due to warp toggling and automatic connection retries rather than a fixed delay.

# Some points to discuss:

## Launch config schemas, compatibility / migration path:

I've kept both the existing `fs-uae` and `winuae` launch config types, but in reality they now map to a single `DebugSession` class. There's really no reason to maintain both other than for backwards compatibility. I was thinking we could implement a new generic 'uae' config type and keep the existing ones for now but declare them as deprecated in the docs and description text. What do you think?

## Binary versions

The adapter will no longer work with the old 3.x binaries. This will obviously result in breakage when people upgrade unless we do something about it. I've done a couple of things to address this. The adapter checks the version string from `fs-uae --version` and returns an error message prompting to upgrade binaries if they need upgrading.

I've decided to go with the option of bundling the patched binaries for all platforms with the `uae-dap` package. My thinking behind this is that they're only ~15mb per platform and this eliminates the headache of having to keep the versions in sync. It's likely that as we implement new features and fixes, the library and binary versions will map 1:1. 

This makes life easier for people using the adapter stand-alone. For this extension we have the option to include the bundled binaries in the package or omit them and stick with the existing pattern if you prefer. I've added a copy command to the webpack config so they're included for now. To use the bundled binary, simply omit the `emulator` property in the launch config. If we are going with the bundling option, perhaps the behavior could be to default to the bundled version if the specified binary path isn't suitable.

I've set up a build process in Github actions so you can always get the latest binaries from the [attached artifacts](https://github.com/grahambates/fs-uae/actions/runs/3082966575) if you want to add them to the pre-compiled binaries package.

prb28 commented:
This is a big step forward!
I totally agree with you with a unified configuration, and to keep but deprecate the old ones.
The "fs-uae --versions" seems to be a good test to warn about the incompatibility.
The bundling seems the better way to do this, I would keep the possibility to set a local path for development or hacking purpose.
Removing the emulator property: how do you choose between winuae and fs-uae on windows ?

grahambates commented:
>  I totally agree with you with a unified configuration, and to keep but deprecate the old ones.

I'll go ahead and implement this. It will basically expose the new `LaunchRequestArguments` from uae-dap.

>  Removing the emulator property: how do you choose between winuae and fs-uae on windows ?

I've added an optional `emulatorType` property in the new config schema which defaults to `winuae` on windows. This controls the default binary path i.e. which bundled emulator to use. This will be exposed once I've added the new unified config type as above.

A couple of other things worth mentioning:
- `LaunchRequestArguments` has a built in property called `noDebug` in the Debug Adapter Protocol. This is now supported so we can use this to provide the 'launch-only' config type rather than needing an additional adapter as we do today.
- The existing `startEmulator` option seems to be better achieved using an attach request vs a launch request. Both are now supported.
- I've implemented logic to automatically add emulator args to configure the remote debugger if they're not already present. This means you no longer need to manually specify `--remote_debugger` etc. This seemed like a good idea because we already know the port number and the name of the program.
